File: src/DSL_model/question.py
from DSL_model import multiple_choice_question, number_question, open_ended_question, single_choice_question, true_false_question

class Question:
  def __init__(self, question_text, question_type, points, negative_points=None):
    self.question_text = question_text
    self.question_type = self.init_type(question_type)
    self.points = points
    self.negative_points = negative_points
    
  def init_type(self, type):
    if type.__class__.__name__ == "MultipleChoiceQuestion":
      return multiple_choice_question.MultipleChoiceQuestion(type.answers, type.accept_partial_answer)
    elif type.__class__.__name__ == "NumberQuestion":
      return number_question.NumberQuestion(type.answer)
    elif type.__class__.__name__ == "OpenEndedQuestion":
      return open_ended_question.OpenEndedQuestion(type.answer)
    elif type.__class__.__name__ == "SingleChoiceQuestion":
      return single_choice_question.SingleChoiceQuestion(type.answers)
    else:
      return true_false_question.TrueFalseQuestion(type.answer)
    
  def to_dict(self):
    return {
        "question_text": self.question_text,
        "question_type": self._type_to_dict(self.question_type),
        "points": self.points,
        "negative_points": self.negative_points
    }
    
  def _type_to_dict(self, question_type):
    if isinstance(question_type, multiple_choice_question.MultipleChoiceQuestion):
        return {
            "type": "MultipleChoiceQuestion",
            "answers": [a.to_dict() for a in question_type.answers],
            "accept_partial_answer": question_type.accept_partial_answer
        }
    elif isinstance(question_type, number_question.NumberQuestion):
        return {
            "type": "NumberQuestion",
            "answer": question_type.answer
        }
    elif isinstance(question_type, open_ended_question.OpenEndedQuestion):
        return {
            "type": "OpenEndedQuestion",
            "answer": question_type.answer.to_dict()
        }
    elif isinstance(question_type, single_choice_question.SingleChoiceQuestion):
        return {
            "type": "SingleChoiceQuestion",
            "answers": [a.to_dict() for a in question_type.answers],
        }
    elif isinstance(question_type, true_false_question.TrueFalseQuestion):
        return {
            "type": "TrueFalseQuestion",
            "answer": question_type.answer
        }
    else:
        raise ValueError("Unknown question type")

File: frontend/curriculum-management/src/components/question/Question.js
import React from "react";
import { useEffect, useState } from "react"
import {
  RadioGroup,
  Radio,
  FormControlLabel,
  FormGroup ,
  Checkbox,
  TextField,
} from "@mui/material";

function Question({ question, onAnswerChange, answer, currentQuestionIndex }) {
  const [loading, setLoading] = useState(true);
  const [shuffledAnswers, setShuffledAnswers] = useState([]);

  useEffect(() => {
    if (question !== undefined) {
      setLoading(false);

      if (question.question_type.type === "SingleChoiceQuestion" || question.question_type.type === "MultipleChoiceQuestion") {
        const answers = question.question_type.answers;
  
        const shuffled = shuffleArray(answers);
        setShuffledAnswers(shuffled);
      }

    }
  }, [question]);

  const shuffleArray = (array) => {
    let shuffled = [...array];
    for (let i = shuffled.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [shuffled[i], shuffled[j]] = [shuffled[j], shuffled[i]];
    }
    return shuffled;
  };

  const renderQuestionInput = () => {
    switch (question.question_type.type) {
      case "OpenEndedQuestion":
        return (
          <TextField
            sx={{
              marginTop: "3%",
              input: { color: "white" }, // Text color
              "& .MuiInputBase-root": {
                color: "white", // Text color
              },
              "& .MuiOutlinedInput-root": {
                "& fieldset": {
                  borderColor: "white", // Border color
                },
                "&:hover fieldset": {
                  borderColor: "white",
                },
                "&.Mui-focused fieldset": {
                  borderColor: "white",
                },
                "&::placeholder": {
                  color: "white", // Placeholder text color
                },
              },
              "& .MuiInputBase-input::placeholder": {
                color: "white", // Additional placeholder text color
                opacity: 0.5, // Ensure full opacity for the placeholder
              },
            }}
            fullWidth
            multiline
            value={answer}
            onChange={(e) => onAnswerChange(e.target.value)}
            variant="outlined"
            placeholder="Type your answer here..."
          />
        );

      case "SingleChoiceQuestion":
        return (
          <RadioGroup
            sx={{
              marginTop: "3%",
              color: "white", // Text color
              "& .MuiRadio-root": {
                color: "white", // Radio button color
              },
            }}
            value={answer}
            onChange={(e) => onAnswerChange(e.target.value)}
          >
            {shuffledAnswers.map((a, index) => (
              <FormControlLabel
                key={index}
                value={a.text}
                control={<Radio />}
                label={a.text}
              />
            ))}
          </RadioGroup>
        );

      case "MultipleChoiceQuestion":
        return (
          <FormGroup>
            {shuffledAnswers.map((a, index) => (
              <FormControlLabel
                sx={{
                  marginTop: "2%",
                  color: "white", // Text color
                  "& .MuiCheckbox-root": {
                    color: "white", // Checkbox color
                  },
                }}
                key={index}
                control={
                  <Checkbox
                    checked={answer.includes(a.text)}
                    onChange={(e) => {
                      const newAnswer = [...answer];
                      if (e.target.checked) {
                        newAnswer.push(a.text);
                      } else {
                        const idx = newAnswer.indexOf(a.text);
                        if (idx > -1) {
                          newAnswer.splice(idx, 1);
                        }
                      }
                      onAnswerChange(newAnswer);
                    }}
                  />
                }
                label={a.text}
              />
            ))}
          </FormGroup>
        );

      case "TrueFalseQuestion":
        return (
          <RadioGroup
            sx={{
              marginTop: "3%",
              color: "white", // Text color
              "& .MuiRadio-root": {
                color: "white", // Radio button color
              },
            }}
            value={answer}
            onChange={(e) => onAnswerChange(e.target.value)}
          >
            <FormControlLabel value="true" control={<Radio />} label="True" />
            <FormControlLabel value="false" control={<Radio />} label="False" />
          </RadioGroup>
        );

      case "NumberQuestion":
        return (
          <TextField
            sx={{
              marginTop: "3%",
              input: { color: "white" }, // Text color
              "& .MuiOutlinedInput-root": {
                "& fieldset": {
                  borderColor: "white", // Border color
                },
                "&:hover fieldset": {
                  borderColor: "white",
                },
                "&.Mui-focused fieldset": {
                  borderColor: "white",
                },
              },
            }}
            type="number"
            value={answer}
            onChange={(e) => onAnswerChange(e.target.value)}
            variant="outlined"
            placeholder="Enter a number"
          />
        );

      default:
        return null;
    }
  };

  if (loading) return <div>Loading...</div>;

  return (
    <div>
      <div>
        Question {currentQuestionIndex + 1}: <b>{question.question_text}</b>
      </div>
      <div style={{ marginTop: "5%" }}>{renderQuestionInput()}</div>
    </div>
  );
}

export default Question;
